daemons/ipa-kdb/ipa_kdb_kdcpolicy.c

```diff
@@ -44,13 +57,112 @@ ipa_kdcpolicy_check_as(krb5_context context, krb5_kdcpolicy_moddata moddata,
                        const char **status, krb5_deltat *lifetime_out,
                        krb5_deltat *renew_lifetime_out)
 {
+    struct ipadb_context *ipactx;
+    LDAPMessage *result = NULL;
+    LDAPMessage *lentry = NULL;
+    krb5_error_code kerr;
+    char *principal = NULL;
+    enum ipadb_user_auth ua;
+    int valid_indicators_set = 0;
+
     *status = NULL;
-    *lifetime_out = 0;
-    *renew_lifetime_out = 0;
 
     krb5_klog_syslog(LOG_INFO, "IPA kdcpolicy: checking AS-REQ.");
 
-    return 0;
+    ipactx = ipadb_get_context(context);
+    if (!ipactx) {
+        kerr = KRB5_KDB_DBNOTINITED;
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: Database not initiallized.");
+        goto done;
+    }
+
+    kerr = krb5_unparse_name(context, request->client, &principal);
+    if (kerr != 0) {
+        krb5_klog_syslog(LOG_ERR, "IPA kdcpolicy: krb5_unparse_name failed.");
+        goto done;
+    }
+
+    /* Filter out anonymous PKINIT principal since it doesn't have a LDAP entry */
+    if (strncmp(principal, "WELLKNOWN/ANONYMOUS@WELLKNOWN:ANONYMOUS", 40) == 0) {
+        goto done;
+    }
+
+    /* Get LDAP entry for the principal */
+    kerr = ipadb_fetch_principals(ipactx, KRB5_KDB_FLAG_ALIAS_OK, principal, &result);
```

Comment:
before going down the path of searching LDAP once more which will considerably slow down the KDC you should check if request->client->e_data does not already contain a struct ipadb_e_data
If you need more data in there please change the default fetching function and store it in that structure so it is readily available without extra searches.

Note that e_Data may not always be available as sometimes the KDC may be constructing the client data instead of getting it from a KDC search, but it is worth checking if e_data is there because it will speed up the operation considerably.

Reply:
Put user_auth flags in e_data and using that now